database2/ffboltdb/db.go

```diff
@@ -0,0 +1,1495 @@
+	checksum := crc32.Checksum(serializedRow[:checksumOffset], castagnoli)
+	byteOrder.PutUint32(serializedRow[checksumOffset:], checksum)
+	return serializedRow
+}
+
+// writePendingAndCommit writes pending block data to the flat block files,
+// updates the metadata with their locations as well as the new current write
+// location, and commits the metadata to the underlying bolt database.  It also
+// properly handles rollback in the case of failures.
+//
+// This function MUST only be called when there is pending data to be written.
+func (tx *transaction) writePendingAndCommit() error {
+	// Save the current block store write position for potential rollback.
+	// These variables are only updated here in this function and there can
+	// only be one write transaction active at a time, so it's safe to store
+	// them for potential rollback.
+	wc := tx.db.store.writeCursor
+	wc.RLock()
+	oldBlkFileNum := wc.curFileNum
+	oldBlkOffset := wc.curOffset
+	wc.RUnlock()
+
+	// rollback is a closure that is used to rollback all writes to the
+	// block files.  It also optionally rolls back the underlying bolt
+	// transaction.
+	rollback := func(rollbackBolt bool) {
+		// Rollback any modification made to the block files and the
+		// underlying bolt transaction if needed.
+		tx.db.store.handleRollback(oldBlkFileNum, oldBlkOffset)
+		if rollbackBolt {
+			_ = tx.boltTx.Rollback()
+		}
+	}
+
+	// Loop through all of the pending blocks to store and write them.
+	for _, blockData := range tx.pendingBlockData {
+		log.Tracef("Storing block %s", blockData.hash)
+		location, err := tx.db.store.writeBlock(blockData.bytes)
+		if err != nil {
+			rollback(true)
+			return err
+		}
+
+		// Add a record in the block index for the block.  The record
+		// includes the location information needed to locate the block
+		// on the filesystem as well as the block header since they are
+		// so commonly needed.
+		blockHdr := blockData.bytes[0:blockHdrSize]
+		blockRow := serializeBlockRow(location, blockHdr)
+		err = tx.blockIdxBucket.Put(blockData.hash[:], blockRow)
+		if err != nil {
+			rollback(true)
+			return err
+		}
+	}
+
+	// Update the metadata for the current write file and offset.
+	writeRow := serializeWriteRow(wc.curFileNum, wc.curOffset)
+	if err := tx.metaBucket.Put(writeLocKeyName, writeRow); err != nil {
+		rollback(true)
+		return convertErr("failed to store write cursor", err)
+	}
+
+	// Commit metadata updates.
+	if err := tx.boltTx.Commit(); err != nil {
+		rollback(false)
+		return convertErr("failed to commit transaction", err)
+	}
+
+	return nil
+}
+
+// rollback rollsback the underly bolt database and closes the transaction.  It
+// is separated mainly so the code panics on attempts to commit or rollback a
+// managed transaction can rollback first.
+func (tx *transaction) rollback() error {
+	// Regardless of whether the rollback succeeds, the transaction is
+	// closed on return.
+	defer tx.close()
```

Comment:
It would need to be in 2 places without the defer.  I can do that if you prefer.